server/src/main/java/io/crate/expression/operator/EqOperator.java

```diff
@@ -189,20 +192,29 @@ public static Query termsQuery(String column, DataType<?> type, Collection<?> va
                 column,
                 nonNullValues.stream().map(x -> new BytesRef(((BitString) x).bitSet().toByteArray())).toList()
             );
-            default -> booleanShould(column, type, nonNullValues);
+            default -> booleanShould(column, type, nonNullValues, hasDocValues, indexType);
         };
     }
 
     @Nullable
-    private static Query booleanShould(String column, DataType<?> type, Collection<?> values) {
+    private static Query booleanShould(String column,
+                                       DataType<?> type,
+                                       Collection<?> values,
+                                       boolean hasDocValues,
+                                       IndexType indexType) {
         BooleanQuery.Builder builder = new BooleanQuery.Builder();
         for (var term : values) {
-            builder.add(EqOperator.fromPrimitive(type, column, term), Occur.SHOULD);
+            builder.add(EqOperator.fromPrimitive(type, column, term, hasDocValues, indexType), Occur.SHOULD);
         }
         return new ConstantScoreQuery(builder.build());
     }
 
-    private static Query termsAndGenericFilter(Function function, String column, DataType<?> elementType, Collection<?> values, LuceneQueryBuilder.Context context) {
+    private static Query termsAndGenericFilter(Function function,
```

Comment:
Why do we need to add `hasDocValues` & `indexType` here? Aren't both false in order to call this function?

Reply:
Thanks for pointing out, have focused so far on `EqQuery` implementations which uses those flags, will look at the execution paths to it.